sympy/functions/combinatorial/factorials.py

```diff
@@ -480,37 +481,38 @@ def eval(cls, x, k):
                         else:
                             return S.Infinity
                     else:
-                        try:
-                            F, opt = poly_from_expr(x)
-                        except PolificationFailed:
-                            return reduce(lambda r, i: r*(x + i), range(0, int(k)), 1)
-                        if len(opt.gens) > 1 or F.degree() <= 1:
-                            return reduce(lambda r, i: r*(x + i), range(0, int(k)), 1)
+                        if isinstance(x, Poly):
+                            gens = x.gens
+                            if len(gens) > 1 or x.degree() <= 1:
```

Comment:
I'm not sure that these cases need to be singled out. I think that this should be called only with univariate polynomials, multivariate being considered as errors. The linear case (`a*y + b, a != 0`) could be handled together with the nonlinear polynomials. Constant polynomials may not give what had been expected (a power instead of a factorial), but those are not strictly speaking univariate either.